state-chain/pallets/cf-elections/src/electoral_systems/blockchain/delta_based_ingress.rs

```diff
@@ -204,102 +204,137 @@ where
+								);
+
+								if new_consensus.amount > old_consensus.amount {
+									(Some(old_consensus), Some(new_consensus))
+								} else {
+									log_or_panic!(
+										"Consensus {:?} balance for Solana deposit channel `{:?}` decreased from {:?} to {:?}",
+										details.asset,
+										account,
+										old_consensus.amount,
+										new_consensus.amount
+									);
+									(Some(old_consensus), None)
+								}
+							} else {
+								// Chain tracking has not progressed beyond the previous deposit
+								// block. We can confirm neither the previous nor the latest amount.
+								(None, Some(new_consensus))
 							}
 						}
 					},
 				};
 
-				if let Some(ingress_total) = option_ingress_total_before_chain_tracking {
+				if let Some(ready_total) = ready_total {
 					let previous_amount = ElectoralAccess::unsynchronised_state_map(&(
 						account.clone(),
 						details.asset,
 					))?
 					.map_or(Zero::zero(), |previous_total_ingressed| {
 						previous_total_ingressed.amount
 					});
-					match previous_amount.cmp(&ingress_total.amount) {
+					match previous_amount.cmp(&ready_total.amount) {
 						Ordering::Less => {
 							Sink::on_ingress(
 								account.clone(),
 								details.asset,
-								ingress_total.amount - previous_amount,
-								ingress_total.block_number,
+								ready_total.amount - previous_amount,
+								ready_total.block_number,
 								(),
 							);
 							ElectoralAccess::set_unsynchronised_state_map(
 								(account.clone(), details.asset),
-								Some(ingress_total),
+								Some(ready_total),
 							);
+							if future_total.is_none() {
+								new_properties.entry(account.clone()).and_modify(
+									|(_details, total)| {
+										*total = ready_total;
+									},
+								);
+							}
 						},
 						Ordering::Greater => {
-							log::warn!("Deposit channels on Solana chain has reverted! Account: {:?}, Asset: {:?}, Previous ingressed total: {:?}, new ingressed total: {:?}", account, details.asset, previous_amount, ingress_total.amount);
+							log::error!(
+								"Finalized {:?} balance for Solana deposit channel `{:?}` decreased from {:?} to {:?}",
+								details.asset,
+								account,
+								previous_amount,
+								ready_total.amount
+							);
 						},
 						Ordering::Equal => (),
 					}
-					if ingress_total.block_number >= details.close_block {
+					if ready_total.block_number >= details.close_block && future_total.is_none() {
 						Sink::on_channel_closed(account.clone());
-						closed_channels.push(account.clone());
+						new_properties.remove(account);
 					}
 				}
-				if let Some(ingress_total_after_chain_tracking) =
-					option_ingress_total_after_chain_tracking
-				{
-					pending_ingress_totals
-						.insert(account.clone(), ingress_total_after_chain_tracking);
+				if let Some(future_total) = future_total {
+					pending_ingress_totals.insert(account.clone(), future_total);
```

Comment:
We are iterating over all channels in `properties`, and for each channel, removing it from `pending_ingress_totals`. Except if we hit this line, then we reinsert it with the new future_total, right?
Since we are anyways removing every entry of `pending_ingress_totals`, and reinserting all that need to be reinserted, couldn't one make `pending_ingress_totals` immutable, and have a `new_pending_ingress_totals`?

I think it works as intended, but keeping it mutable suggests there is more going on then there actually is

Reply:
Did this.